### PR TITLE
Resize 2D and 4D arrays with Fourier interpolation

### PR DESCRIPTION
**Pull request by cophus**
This update adds a single function to utils which can change the size of images / 4D datasets using Fourier inteprolation.  Also includes band limit and positivity constraints as optional inputs.

### CODE REVIEW

**Review by bsavitzky — CHANGES_REQUESTED** [review]
Submitting this review of the docstring without reading the source code yet, as @sezelt is going to be doing a refactor of the code first

**Comment by cophus** [comment]
I accepted all changes from @sezelt  

@bsavitzky this is ready for another review

File: py4DSTEM/process/utils/utils.py [hunk 1]
```diff
@@ -607,3 +608,204 @@ def get_ewpc_filter_function(Q_Nx, Q_Ny):
...
+    scale_output = np.prod(output_size) / np.prod(input__size)
+
+
+    if bandlimit_nyquist is not None:
+        if len(input__size) == 2:
+            kx = np.fft.fftfreq(output_size[0])
+            ky = np.fft.fftfreq(output_size[1])
+        if len(input__size) == 4:
+            kx = np.fft.fftfreq(output_size[2])
+            ky = np.fft.fftfreq(output_size[3])
```
Comment (bsavitzky) [inline py4DSTEM/process/utils/utils.py:664]: I'm pretty sure this if statement isn't doing anything....the way the logic above is set up, both `input__size` and `output__size` will always have 2 elements.

File: py4DSTEM/process/utils/utils.py [hunk 2]
```diff
@@ -607,3 +608,204 @@ def get_ewpc_filter_function(Q_Nx, Q_Ny):
...
+        x_lr_out = slice(0-input__size[0]//2+output_size[0], output_size[0])
+        x_lr_in_ = slice(0-input__size[0]//2+input__size[0], input__size[0])
+
+    elif output_size[0] < input__size[0]:
+        # x dimension decreases
+        x_ul_out = slice(0, output_size[0]//2)
+        x_ul_in_ = slice(0, output_size[0]//2)
+
+        x_ll_out = slice(0-output_size[0]//2, output_size[0])
+        x_ll_in_ = slice(0-output_size[0]//2+input__size[0], input__size[0])
```
Comment (bsavitzky) [inline py4DSTEM/process/utils/utils.py:695]: I'm not sure what this code is doing exactly yet, but symmetry considerations suggest to me that there may be something wrong here...?

File: py4DSTEM/process/utils/utils.py [hunk 3]
```diff
@@ -607,3 +608,204 @@ def get_ewpc_filter_function(Q_Nx, Q_Ny):
...
+
+        x_ur_out = slice(0, output_size[0]//2)
+        x_ur_in_ = slice(0, output_size[0]//2)
+
+        x_lr_out = slice(0-output_size[0]//2+output_size[0], output_size[0])
+        x_lr_in_ = slice(0-output_size[0]//2+input__size[0], input__size[0])
```
Comment (bsavitzky) [inline py4DSTEM/process/utils/utils.py:701]: or, alternatively, here.....?

**Review by bsavitzky — CHANGES_REQUESTED** [review]
Please see comments! This will be excellent new functionality to have in utils :D

File: py4DSTEM/process/utils/utils.py (hunk 1 above)
Comment (sezelt) [reply py4DSTEM/process/utils/utils.py:664]: good catch, this should be `if np.size(array.shape) == 4:`

Comment (bsavitzky) [reply py4DSTEM/process/utils/utils.py:664]: I think you still don't actually need the if statements checking the length at all.  The way you've written this, output_size should always have length 2, and so regardless of whether we're dealing with 2 or 4D input arrays, you should just be able to use
```
kx = np.fft.fftfreq(output_size[0])
ky = np.fft.fftfreq(output_size[1])
```

File: py4DSTEM/process/utils/utils.py [hunk 4]
```diff
@@ -607,3 +608,204 @@ def get_ewpc_filter_function(Q_Nx, Q_Ny):
...
+    bandlimit_power=2, 
+    dtype=np.float32):
+    """
+    Resize a 2D array along any dimension, using Fourier interpolation / extrapolation.
+    For 4D input arrays, only the final two axes can be resized.
+
+    The scaling of the array can be specified by passing either `scale`, which sets
+        the scaling factor along both axes to be scaled; or by passing `output_size`, 
+        which specifies the final dimensions of the scaled axes (and allows for different
+        scaling along the x,y or kx,ky axes.)
```
[the remark this answers is not in the document]
Comment (cophus) [reply py4DSTEM/process/utils/utils.py:628]: Habit?   Haha.

likely I just shift-tabbed without even thinking about it.

File: py4DSTEM/process/utils/utils.py (hunk 1 above)
Comment (sezelt) [reply py4DSTEM/process/utils/utils.py:664]: aha even better, i ingeniously placed the buggy code in an inaccessible branch

File: py4DSTEM/process/utils/utils.py [hunk 5]
```diff
@@ -607,3 +608,204 @@ def get_ewpc_filter_function(Q_Nx, Q_Ny):
...
+        output_size (2-tuple of ints): two values giving either the (x,y) output size for 2D, or (kx,ky) for 4D
+        force_nonnegative (bool): Force all outputs to be nonnegative, after filtering
+        bandlimit_nyquist (float): Gaussian filter information limit in Nyquist units (0.5 max in both directions)
+        bandlimit_power (float): Gaussian filter power law scaling (higher is sharper)
+        dtype (numpy dtype): datatype for binned array. default is single precision float
+
+    Returns:
+        the resized array (2D/4D numpy array)
+    """
+    input__size = array.shape[-2:]
```
[the remark this answers is not in the document]
Comment (cophus) [reply py4DSTEM/process/utils/utils.py:642]: Added with a helpful exception statement:
    **# Verify input is 2D or 4D
    if np.size(array.shape) != 2 and np.size(array.shape) != 4:
        raise Exception('Function does not support arrays with ' \
         + str(np.size(array.shape)) + ' dimensions')**

File: py4DSTEM/process/utils/utils.py (hunk 1 above)
Comment (cophus) [reply py4DSTEM/process/utils/utils.py:664]: Correctomundo!

File: py4DSTEM/process/utils/utils.py (hunk 2 above)
Comment (cophus) [reply py4DSTEM/process/utils/utils.py:695]: Tested and outputs seem correct!  I believe it matches my original long-form written out version.

Comment (cophus) [reply py4DSTEM/process/utils/utils.py:695]: With further reflection, I actually think you are correct @bsavitzky !

I am working on simplifying it further.

Comment (cophus) [reply py4DSTEM/process/utils/utils.py:695]: New version fixes the half spaces for even and odd, and repairs the symmetry.  Noether strikes again!

**Comment by cophus** [comment]
Ready for re-review!

**Review by bsavitzky — APPROVED** [review]

**Pull request merged**